Questions 1-49 are based on the Amazon Book Dataset. 
Questions 50-108 use artificial data on educational website, structured across six tables: 
Users, Courses, Ratings, Payments, Enrollments, and Refunds ! 

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime, timedelta

In [3]:
df = pd.read_csv("C:\\Users\\HP\\OneDrive\\Attachments\\Book_store.csv")
df

Book Name                Author  \
0               11 Rules For Life: Secrets to Level Up         Chetan Bhagat   
1                       1984 [Paperback] George Orwell         George Orwell   
2                       1984 [Paperback] George Orwell         George Orwell   
3    21 Lessons for the 21st Century : 'Truly mind-...     Yuval Noah Harari   
4    21 Lessons for the 21st Century : 'Truly mind-...     Yuval Noah Harari   
..                                                 ...                   ...   
315           You Can [Paperback] George Matthew Adams  George Matthew Adams   
316  You Can by George Matthew Adams – Tamil Transl...  George Matthew Adams   
317               You Got This: Art of Loving Yourself           Gianna Mago   
318  YOU WILL GET THROUGH THIS NIGHT NOTUS NO: The ...         DANIEL HOWELL   
319                                    Zero to One (L)           Peter Thiel   

     Pages Language  Ratings Total Ratings Price                  Category  
0    256.0  English      4.5           735   183          Self Improvement  
1    304.0  English      4.5      1,26,238   139                Literature  
2    304.0  English      4.5      1,26,238   139                Literature  
3    432.0  English      4.6        17,463   416          Self Improvement  
4    432.0  English      4.6        17,463   416          Self Improvement  
..     ...      ...      ...           ...   ...                       ...  
315  232.0  English      4.3         5,336   129          Self Improvement  
316  224.0    Tamil      4.2            37   129                   Finance  
317  128.0  English      4.6            16   205          Self Improvement  
318  352.0  English      4.9         2,258   860  Religion and Spirtuality  
319  210.0  English      4.5        30,885   381                   Finance  

[320 rows x 8 columns]

**1. Get the names of all the columns**

In [5]:
df.columns

Index(['Book Name', 'Author', 'Pages', 'Language', 'Ratings', 'Total Ratings',
       'Price', 'Category'],
      dtype='object')

**2. Display the first 3 rows**

In [7]:
df.head(3)

Book Name         Author  Pages Language  \
0  11 Rules For Life: Secrets to Level Up  Chetan Bhagat  256.0  English   
1          1984 [Paperback] George Orwell  George Orwell  304.0  English   
2          1984 [Paperback] George Orwell  George Orwell  304.0  English   

   Ratings Total Ratings Price          Category  
0      4.5           735   183  Self Improvement  
1      4.5      1,26,238   139        Literature  
2      4.5      1,26,238   139        Literature

**3. Display 4 random rows from the dataframe**

In [9]:
df.loc[4],[8],[10]

(Book Name        21 Lessons for the 21st Century : 'Truly mind-...
 Author                                           Yuval Noah Harari
 Pages                                                        432.0
 Language                                                   English
 Ratings                                                        4.6
 Total Ratings                                               17,463
 Price                                                          416
 Category                                          Self Improvement
 Name: 4, dtype: object,
 [8],
 [10])

**4. What is the total number of rows in the dataset?**

In [11]:
total_rows = df.shape
total_rows

(320, 8)

In [12]:
## rename ratings to average_ratings

In [13]:
renaming = df.rename(columns = {'Ratings':'Average_Rating'},inplace =True)
renaming

**6. Check if there are any duplicate rows**

In [15]:
df.duplicated()

0      False
1      False
2       True
3      False
4       True
       ...  
315     True
316    False
317    False
318    False
319    False
Length: 320, dtype: bool

**7. Remove any duplicate rows**

In [17]:
remove_duplicates = df.drop_duplicates(inplace = True)
remove_duplicates

**8. Remove rows where 'Average Rating' and 'Total Ratings' are missing**

In [19]:
df.dropna(subset = ['Average_Rating','Total Ratings'], inplace = True)

In [41]:
#removve = df.dropna(subset = ['Ratings'])
df.isnull().sum()


Book Name          0
Author             0
Pages              0
Language           0
Average_Rating     0
Total Ratings      0
Price             19
Category           0
dtype: int64

**9. Check the datatype of all the columns**

In [43]:
df.dtypes

Book Name          object
Author             object
Pages               int32
Language           object
Average_Rating    float64
Total Ratings      object
Price              object
Category           object
dtype: object

**10. Remove comma from values in Price column and change the type to float64**

In [45]:
#df['Price'] = df['Price'].str.replace(',','',regex =True).astype(float)
#df['Price']
#df['Price'] = pd.to_numeric(df['Price'],errors = 'coerce')
df.dtypes

Book Name          object
Author             object
Pages               int32
Language           object
Average_Rating    float64
Total Ratings      object
Price              object
Category           object
dtype: object

**11. Remove comma from values in Total Ratings column and change the type to int64**

In [26]:
df['Total Ratings'] = df['Total Ratings'].str.replace(',','',regex = True)
df['Total Ratings']


0         735
1      126238
3       17463
5        3391
7        2536
        ...  
314      5336
316        37
317        16
318      2258
319     30885
Name: Total Ratings, Length: 239, dtype: object

**12. Check how many missing values are there in all columns**

In [28]:
df.isnull().sum()

Book Name          0
Author             2
Pages             24
Language           5
Average_Rating     0
Total Ratings      0
Price             19
Category           0
dtype: int64

**13. Replace missing values in Author column with 'Unknown'**

In [30]:
df['Author']= df['Author'].fillna('Unknown')

**14. Replace missing values in Language column with the most occuring langauge**

In [32]:
#df[df['Language'].isnull()]
df['Language'] = df['Language'].fillna(df['Language'].mode()[0])


**15. Replace missing values in Pages column with median and convert data of Pages column type to int64**

In [34]:
df['Pages'] = df['Pages'].fillna(df['Pages'].median()).astype(int)
#df[df['Pages'].isnull()]
#df.isnull().sum()


**16. Fill missing values for price with the median price of the same category**

In [36]:

#df['Price'] = pd.to_numeric(df['Price'],errors ='coerce')
#df.dtypes
df['Price'] = df.groupby('Category')['Price'].transform(lambda x : x.fillna(x.median()))

TypeError: Cannot convert ['183' '139' '274' '159' '662' '499' '286' '199' nan nan '199' '215' '125'
 '239' '1,949' '253' '209' '532' nan '153' '635' '393' '139' '259' '111'
 '257' '267' '209' '103' '264' '1,490' '499' '619' '449' '149' '169' '138'
 nan '267' '149' nan '1,155' '260' '149' '129' '381'] to numeric

**17. How many unique authors are listed?**

In [ ]:
unique_Author = df['Author'].unique()
unique_Author

**18. What are the top 5 most expensive books?**

In [ ]:
#df.sort_values('Price',ascending = False).head(5)
df.nlargest(5,'Price')['Book Name']

**19. What are the 5 shortest books (based on the number of pages)?**  


In [ ]:
df.nsmallest(5, 'Pages')[['Book Name','Pages']]

**20. What is the average number of pages across all books?**


In [ ]:
df['Pages'].mean()

**21. Display the summary statistics for numeric columns**



In [ ]:
df.describe()

**22. Find the unique values and their counts for the column 'Category', sort by category names**

In [ ]:
df.groupby('Category').count().sort_values('Category')

**23. What are the top 5 languages with the most books listed?**


In [ ]:
df.groupby('Language').count().nlargest(5, 'Book Name')

**24. What percentage of books have ratings greater than or equal to 4 stars?**

In [ ]:
df.groupby('Total Ratings')

**25. What is the average price of books and average number of pages across different languages?**

In [ ]:
avg_price_pages = df[['Price', 'Pages']].mean()
avg_price_pages

**26. What is the average and maximum price of books, along with the average number of pages, across different languages?**

In [39]:
avg_max_price = df[['Price', 'Pages']].agg(['mean', 'max'])
avg_max_price


TypeError: can only concatenate str (not "int") to str

**27. How do the average price and average number of pages differ across various languages and categories**

In [37]:
df['Price'].mean()

TypeError: can only concatenate str (not "int") to str

**28. Extract the data for the 'English' language from the previously grouped DataFrame.**

**29. Extract the data for the combination of 'English' language and 'Finance' category from the previously grouped DataFrame.**

**30. Which price range has the most number of books?**

**31. How many books have received more than 1000 ratings?**


In [ ]:
g = df.groupby('Book Name')
g
g.sum().sort_values('Total Ratings')

**32. Which book(s) has the highest rating?**


**33. What is the average rating across different languages?**

**34. Which authors have the highest average book ratings?**


**35. Display the longest book(max pages) written by each author.  Sort the result by Pages in descending order**

**36. Display the total number of ratings received by each author.**



**37. Find top 5 and bottom 5 authors based on total number of ratings received**

**38. Find the top 10 books with the highest number of total ratings.**


**39. Explore how many books exist in each category and language combination.**

**40 Identify the longest and shortest books by pages within each category.**

**41. Plot a boxplot of the "Pages" column**

**42. Plot a boxplot of the "Price" column**

**43. Use the following function to remove outliers from Price and Pages column. Find the number of rows removed and plot boxplots for Price and Pages after removing the oultiers**

In [ ]:
def remove_outliers(df, column, multiplier=1.5):
    # Calculate Q1 (25th percentile) and Q3 (75th percentile) of the column
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)

    # Calculate IQR
    IQR = Q3 - Q1

    # Define the lower and upper bounds for outliers
    lower_bound = Q1 - multiplier * IQR
    upper_bound = Q3 + multiplier * IQR

    # Remove outliers by filtering the column within the bounds
    df_cleaned = df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]

    return df_cleaned

**44. Compare the distribution of book prices across different categories using box plots.**


**45. Explore how the average rating varies across different languages using a bar plot**


**46. Analyze the correlation between numerical variables such as Price, Pages, Average Rating, Total Ratings, etc.**

**47. Visualize the correlation between book price and number of pages.**

**48. Plot a histogram for the distribution of ratings**

**49. Draw a line plot to show average price for each rating level**

# Educational Website 

In [ ]:
# Read all CSV files into DataFrames
users_df = pd.read_csv("users.csv",parse_dates=['date_of_birth','registration_date'], dayfirst=True)
courses_df = pd.read_csv("courses.csv",parse_dates=['created_on','last_updated_on'], dayfirst=True)
ratings_df = pd.read_csv("ratings.csv",parse_dates=['first_reviewed_on','last_updated_on'], dayfirst=True)
enrollments_df = pd.read_csv("enrollments.csv",parse_dates=['enrollment_date'], dayfirst=True)
payments_df = pd.read_csv("payments.csv",parse_dates=['transaction_date'], dayfirst=True)
refunds_df = pd.read_csv("refunds.csv",parse_dates=['refund_date'], dayfirst=True)

In [ ]:
print("Users DataFrame columns:", users_df.columns.tolist(), "\n")
print("Courses DataFrame columns:", courses_df.columns.tolist(), "\n")
print("Ratings DataFrame columns:", ratings_df.columns.tolist(), "\n")
print("Enrollments DataFrame columns:", enrollments_df.columns.tolist(), "\n")
print("Payments DataFrame columns:", payments_df.columns.tolist(), "\n")
print("Refunds DataFrame columns:", refunds_df.columns.tolist(), "\n")

**50. How many courses are free and how many courses have a price associated with them?**


**51. How many courses is each instructor teaching on the platform?**

**52. Which courses are the most popular based on the number of bookmarks?**

**53. How many users are registered in each user category (e.g., Working, Student, Homemaker, Retired)?**

**54. What is the male-to-female ratio among the users?**

**55. Which courses have the worst completion rates?**


**56. What is the most common rating given by students?**

**57. How many unique users are enrolled in at least one course?**

**58. How many users are enrolled in multiple courses?**

**59. Which 5 courses have the most enrollments?**

**60. Find the total revenue generated from successful payments.**

**61. What is the average discount percentage applied across all discounted payments?**

**62. Find top 3 countries with the highest number of students**

**63. What is the most common reasons for refunds?**

**64. Add an age column to the users table and then calculate average age of the students on the website**

**65. Find average age of the students in each course**

**66. Find average age of the students in each course category**

**67. Calculate the average age of users for each user category (e.g., Working, Student, Homemaker, Retired)**


**68. Plot the age distribution of all the users**

**69. Plot the gender distribution to understand the gender breakdown of users**

**70. Plot the geographic distribution of users (based on country)**

**71. Show how the age distribution varies for different user types.**

**72. Find the top 5 most popular courses based on view counts**

**73. Find the top 2 popular courses in each category based on view counts**

**74. What is the average course duration per category.**

**75. Find the top 5 highest revenue generating courses**

**76. What is the average transaction amount for each payment method.**

**77. Calulate the average rating for each course and add a column average_rating in courses_df**

**78. Find 2 top rated courses in each course category**

**79. Display the average rating of each course category**

**80. Calculate the average rating for each instructor.**


**81. Display courses with no ratings**

**82. Find courses that have no students enrolled in them**

**83. Find the number of students from each country based on their enrollments.**


**84. Get the course popularity details by country i.e. find the course with the highest number of students in each country.**

**85. Find the top 5 bestselling paid courses**

**86. Find the bestselling course in each category**

**87. Calculate the number of free enrollments and paid enrollments**

**88. Find correlation between the number of videos and the completion_rate.**

**89. Find the correlation between number of Quizzes and average rating for each course**

**90. How many payment transactions failed using paypal**

**91. Get a summary of payment methods, including how many successful, failed or pending payments are there for each method.**

**92. Display the courses that were never refunded**

**93. Use box plot to plot distribution of completion rates by difficulty level.**

**94. Identify the course category with the most enrollments**

**95. Find the number of inactive users - users who have not logged in during the last 30 days.**

**96. What is the average price of courses in each category? (exclude the free courses)**

**97. How many courses are available with subtitles?**

**98. What percentage of courses offer certifications?**

**99. What is the average time gap between a user's registration date and their first enrollment date?**

**100. Get the email addresses and names of users who have signed up but have not enrolled in any course.**

**101. Find the customers who are refunding without any reason**

**102. Analyze monthly new user growth to understand monthly trends**

**103. Calculate and plot the cumulative user growth**

**104. Generate the monthly revenue report** 

**105. Generate the annual revenue report** 

**106. Analyze whether there are more purchases on weekdays or weekends**

**107. Which coupon code is used most frequently?  Visualize the distribution of coupon code usage**

**108. Analyze completion rates by country**